tests/test_clustering.py

```diff
@@ -0,0 +1,209 @@
+
+    class DummyClusterer:
+
+        def __init__(self):
+            pass
+
+        def fit(self):
+            pass
+
+        def fit_predict(self):
+            pass
+
+    clf = DummyClusterer()
+    with pytest.raises(TypeError):
+        plot.elbow_curve(X, clf)
+
+
+def test_cluster_ranges():
+    clf = KMeans()
+    plot.elbow_curve(X, clf, n_clusters=range(1, 10))
+
+
+@image_comparison(baseline_images=['elbow_curve_from_results'],
+                  extensions=['png'],
+                  remove_text=False)
+def test_elbow_curve_from_results():
+    n_clusters = range(1, 10, 2)
+    sum_of_squares = np.array([4572.2, 470.7, 389.9, 335.1, 305.5])
+    plot.elbow_curve_from_results(n_clusters, sum_of_squares, times=None)
+
+
+@image_comparison(baseline_images=['elbow_curve_from_results'],
+                  extensions=['png'],
+                  remove_text=False)
+def test_elbow_curve_from_results_unsorted():
+    n_clusters = [5, 3, 9, 1, 7]
+    sum_of_squares = np.array([389.9, 470.7, 305.5, 4572.2, 335.1])
+    plot.elbow_curve_from_results(n_clusters, sum_of_squares, times=None)
+
+
+def test_ax_elbow():
+    clf = KMeans()
+    fig, ax = plt.subplots(1, 1)
+    out_ax = plot.elbow_curve(X, clf, ax=ax)
+    assert ax is out_ax
+
+
+def test_n_jobs():
+    clf = KMeans()
+    plot.elbow_curve(X, clf, n_jobs=2)
+
+
+@image_comparison(baseline_images=[
+    'silhouette_plot_two_clusters', 'silhouette_plot_three_clusters',
+    'silhouette_plot_four_clusters', 'silhouette_plot_five_clusters',
+    'silhouette_plot_six_clusters'
+],
+                  extensions=['png'],
+                  remove_text=False)
+def test_plot_silhouette():
+    clf = KMeans()
+    plot.silhouette_plot(X, clf)
+
+
+@image_comparison(baseline_images=[
+    'silhouette_plot_four_clusters', 'silhouette_plot_five_clusters'
+],
+                  extensions=['png'],
+                  remove_text=False)
+def test_plot_silhouette_with_cluster_range():
+    clf = KMeans(random_state=10)
+    plot.silhouette_plot(X, clf, range_n_clusters=[4, 5])
+
+
+@image_comparison(baseline_images=[
+    'silhouette_plot_four_clusters_minibatchkmeans',
+    'silhouette_plot_five_clusters_minibatchkmeans'
+],
+                  extensions=['png'],
+                  remove_text=False)
+def test_plot_silhouette_with_minibatchkmeans():
+    clf = MiniBatchKMeans(random_state=10)
+    plot.silhouette_plot(X, clf, range_n_clusters=[4, 5])
+
+
+@image_comparison(baseline_images=['silhouette_plot_spectral'],
+                  extensions=['png'],
+                  remove_text=False)
+def test_cmap():
+    clf = KMeans()
+    plot.silhouette_plot(X, clf, range_n_clusters=[2], cmap='Spectral')
+
+
+@image_comparison(baseline_images=['silhouette_plot_cosine'],
+                  extensions=['png'],
+                  remove_text=False)
+def test_metric():
+    clf = KMeans()
+    plot.silhouette_plot(X, clf, range_n_clusters=[6], metric='cosine')
+
+
+def test_string_classes():
```

Comment:
This test generates the file `silhouette_plot_string_classes.png` in the root folder. I suggest saving this file in some dedicated directory for test artifacts that we can add to .gitignore.

Reply:
This was a miss from my end. I have removed the saving logic.